src/MOIWrapper.jl

```diff
@@ -0,0 +1,373 @@
+            Dict{Int, UnitRange{Int}}())
+    end
+end
+
+mutable struct SCSOptimizer <: MOI.AbstractOptimizer
+    cone::Cone
+    maxsense::Bool
+    data::Union{Void, Data} # only non-Void between MOI.copy! and MOI.optimize!
+    sol::MOISolution
+    function SCSOptimizer()
+        new(Cone(), false, nothing, MOISolution())
+    end
+end
+
+function MOI.isempty(optimizer::SCSOptimizer)
+    !optimizer.maxsense && optimizer.data === nothing
+end
+function MOI.empty!(optimizer::SCSOptimizer)
+    optimizer.maxsense = false
+    optimizer.data = nothing # It should already be nothing except if an error is thrown inside copy!
+end
+
+MOI.canaddvariable(optimizer::SCSOptimizer) = false
+
+MOI.supports(::SCSOptimizer, ::MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}) = true
+MOI.supportsconstraint(::SCSOptimizer, ::Type{<:SF}, ::Type{<:SS}) = true
+MOI.copy!(dest::SCSOptimizer, src::MOI.ModelLike; copynames=true) = MOIU.allocateload!(dest, src, copynames)
+
+# Implements optimize! : translate data to SCSData and call SCS_solve
+using Compat.SparseArrays
+
+const ZeroCones = Union{MOI.EqualTo, MOI.Zeros}
+const LPCones = Union{MOI.GreaterThan, MOI.LessThan, MOI.Nonnegatives, MOI.Nonpositives}
+
+_dim(s::MOI.AbstractScalarSet) = 1
+_dim(s::MOI.AbstractVectorSet) = MOI.dimension(s)
+
+# Computes cone dimensions
+constroffset(cone::Cone, ci::CI{<:MOI.AbstractFunction, <:ZeroCones}) = ci.value
+function _allocateconstraint!(cone::Cone, f, s::ZeroCones)
+    ci = cone.f
+    cone.f += _dim(s)
+    ci
+end
+constroffset(cone::Cone, ci::CI{<:MOI.AbstractFunction, <:LPCones}) = cone.f + ci.value
+function _allocateconstraint!(cone::Cone, f, s::LPCones)
+    ci = cone.l
+    cone.l += _dim(s)
+    ci
+end
+constroffset(cone::Cone, ci::CI{<:MOI.AbstractFunction, <:MOI.SecondOrderCone}) = cone.f + cone.l + ci.value
+function _allocateconstraint!(cone::Cone, f, s::MOI.SecondOrderCone)
+    push!(cone.qa, s.dimension)
+    ci = cone.q
+    cone.q += _dim(s)
+    ci
+end
+constroffset(cone::Cone, ci::CI{<:MOI.AbstractFunction, <:MOI.PositiveSemidefiniteConeTriangle}) = cone.f + cone.l + cone.q + ci.value
+function _allocateconstraint!(cone::Cone, f, s::MOI.PositiveSemidefiniteConeTriangle)
+    push!(cone.sa, s.dimension)
+    ci = cone.s
+    cone.s += _dim(s)
+    ci
+end
+constroffset(cone::Cone, ci::CI{<:MOI.AbstractFunction, <:MOI.ExponentialCone}) = cone.f + cone.l + cone.q + cone.s + ci.value
+function _allocateconstraint!(cone::Cone, f, s::MOI.ExponentialCone)
+    ci = 3cone.ep
+    cone.ep += 1
+    ci
+end
+constroffset(optimizer::SCSOptimizer, ci::CI) = constroffset(optimizer.cone, ci::CI)
+MOIU.canallocateconstraint(::SCSOptimizer, ::Type{<:SF}, ::Type{<:SS}) = true
+function MOIU.allocateconstraint!(optimizer::SCSOptimizer, f::F, s::S) where {F <: MOI.AbstractFunction, S <: MOI.AbstractSet}
+    CI{F, S}(_allocateconstraint!(optimizer.cone, f, s))
+end
+
+# Vectorized length for matrix dimension n
+sympackedlen(n) = (n*(n+1)) >> 1
```

Comment:
Use `div(..., 2)` instead of `... >> 1`. Favor clarity over shortcuts unless there's a good reason.